src/udev/net/ethtool-util.c

```diff
@@ -572,7 +593,15 @@ int ethtool_set_glinksettings(int *fd, const char *ifname, struct link_config *l
         if (link->port != _NET_DEV_PORT_INVALID)
               u->base.port = link->port;
 
-        u->base.autoneg = link->autonegotiation;
+        if (u->base.autoneg != link->autonegotiation)
+                u->base.autoneg = link->autonegotiation;
+
+        if (link->advertise) {
+                uint32_t advertise[ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NU32] = {};
+
+                advertise[0] = link->advertise;
+                memcpy(&u->link_modes.advertising, &advertise, ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NBYTES);
```

Comment:
What? Is it correct?

Reply:
yes same as ethtool.

Reply:
Still I am confused... I do not have enough knowledge about ethtool, but superficially I guess this may be the following...
```
memcpy(u->link_modes.advertising, advertise, ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NBYTES);
```

Reply:
@flo-wer has given a good explanation

Reply:
more for readability

Reply:
I agree with @yuwata here. While the `&` syntax works because `advertise` and `advertising` are arrays (defined with `[`*length*`]`), this wouldn't work if they were pointers instead. I'd prefer that code would work if arrays and pointers would work interchangeably (because a refactor could turn one into the other, with that assumption.) I'd recommend removing the `&`s here, makes it more straightforward.

Reply:
Why it won't work with pointers. Memcpy takes a defined length.

Reply:
Because with a pointer, like `uint32_t *advertise`, `&advertise` refers to the address of the *pointer* itself, not to the buffer it points to...

If you rewrite that code above as:

```c
        uint32_t advertise_buffer[ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NU32] = {};
        uint32_t *advertise = advertise_buffer;
```

The `memcpy` as you currently have will break... If you remove the `&`s, then it will work either way.

Reply:
Precisely why I would pass the pointer address rather than the buffer. Pointer holds the buffer address. Does not makes any sense passing pointer address. Here it's not a pointer its a buffer.

Reply:
BTW 
```
 int a[10] = {};
a, &a,&a[0] mean the same .
````

Reply:
But they have different types...

`a` or `&a[0]` (which are really the same) have a type that matches `int *`.

`&a` has a type that matches `int (*)[10]`.

See [this](https://eli.thegreenplace.net/2010/01/11/pointers-to-arrays-in-c/) for more details.

My initial reaction was the same as @yuwata. This looks fairly unusual to me. I'd prefer to see it without the `&`s.

I think if you look at other places in systemd code (or other code bases) where `memcpy` is used on buffers, I believe you'll find most of them will not use `&`s... (But, admittedly, I haven't looked...)

Reply:
>a or &a[0] (which are really the same) have a type that matches int *.
&a has a type that matches int (*)[10].
 a  is pointer that can point to an array of 10 integers
```

    int arr[] = { 3, 5, 6, 7, 9 };
    int *p = arr;
    int (*ptr)[5] = &arr
```
https://www.geeksforgeeks.org/pointer-array-array-pointer/


>>My initial reaction was the same as @yuwata. This looks fairly unusual to me. I'd prefer to see it without the &s.
I think if you look at other places in systemd code (or other code bases) where memcpy is used on buffers, I believe you'll find most of them will not use &s... (But, admittedly, I haven't looked...)

Just did a quick search. Well it's been used few places src/core/bpf-firewall.c:350 and other places too .

Reply:
> Just did a quick search. Well it's been used few places `src/core/bpf-firewall.c:350` and other places too.

That one is different...

```c
memcpy(key_ipv4->data, &a->address, sizeof(uint32_t));
```

But `a->address` is a `union in_addr_union` which is either a `struct in_addr` or a `struct in6_addr`. In that case, `&` is needed. If you remove it, it doesn't compile:

```
../src/core/bpf-firewall.c:350:48: error: incompatible type for argument 2 of ‘memcpy’
                         memcpy(key_ipv4->data, a->address, sizeof(uint32_t));
                                                ^
In file included from ../src/core/bpf-firewall.c:20:0:
/usr/include/string.h:42:14: note: expected ‘const void * restrict’ but argument is of type ‘union in_addr_union’
 extern void *memcpy (void *__restrict __dest, const void *__restrict __src,
```

Reply:
why ?

Reply:
@filbranden and @yuwata listed the reasons. Also for consistency with the rest of the codebase.

Reply:
>@filbranden and @yuwata listed the reasons. Also for consistency with the rest of the codebase.
No I am not really agree the reason with but yes consistency is ok but still then we have similar code.

Reply:
memcpy just looks for address. so I Do not at all agree with your explanation. Array and pointer to the array at the same location means the same. Not using as coding style is OK